src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -210,34 +201,61 @@ internal bool NullableSemanticAnalysisEnabled
         /// </summary>
         internal bool IsPeVerifyCompatEnabled => LanguageVersion < LanguageVersion.CSharp7_2 || Feature("peverify-compat") != null;
 
-        internal bool ShouldRunNullableWalker
+        /// <summary>
+        /// Returns true if nullable analysis should be run in this compilation.
+        /// </summary>
+        /// <return>
+        /// Returns true if analysis is explicitly enabled for all methods;
+        /// false if analysis is explicitly disabled; and otherwise returns true
+        /// if there are any nullable-enabled contexts in the compilation.
+        /// </return>
+        internal bool ShouldRunNullableAnalysis
         {
             get
             {
-                if (!_lazyShouldRunNullableWalker.HasValue())
+                if (!_lazyShouldRunNullableAnalysis.HasValue())
+                {
+                    _lazyShouldRunNullableAnalysis = (GetNullableAnalysisValue() ?? hasEnabledContexts()).ToThreeState();
+                }
+                return _lazyShouldRunNullableAnalysis.Value();
+
+                bool hasEnabledContexts()
                 {
                     if (Options.NullableContextOptions != NullableContextOptions.Disable)
                     {
-                        _lazyShouldRunNullableWalker = ThreeState.True;
                         return true;
                     }
-
-                    foreach (var syntaxTree in SyntaxTrees)
-                    {
-                        if (((CSharpSyntaxTree)syntaxTree).HasNullableEnables())
-                        {
-                            _lazyShouldRunNullableWalker = ThreeState.True;
-                            return true;
-                        }
-                    }
-
-                    _lazyShouldRunNullableWalker = ThreeState.False;
+                    return SyntaxTrees.Any(tree => ((CSharpSyntaxTree)tree).HasNullableEnables());
                 }
-
-                return _lazyShouldRunNullableWalker.Value();
             }
         }
 
+#if DEBUG
+        /// <summary>
+        /// Returns true if nullable analysis should be run and results ignored in cases where analysis
+        /// is not otherwise necessary. Used for testing in debug builds to increase the chance of catching
+        /// nullable regressions (e.g. https://github.com/dotnet/roslyn/issues/40136).
+        /// </summary>
+        /// <return>
+        /// Returns true unless analysis is explicitly disabled.
+        /// </return>
+        internal bool ShouldRunNullableAnalysisAndIgnoreResults => GetNullableAnalysisValue() != false;
+#endif
+
+        /// <summary>
+        /// Returns Feature("run-nullable-analysis") as a bool? value:
+        /// true for "always"; false for "never"; and null otherwise.
+        /// </summary>
+        private bool? GetNullableAnalysisValue()
+        {
+            return Feature("run-nullable-analysis") switch
```

Comment:
>"run-nullable-analysis" [](start = 27, length = 23)

It looks like at least one IDE test in CSharpCompletionCommandHandlerTests.vb takes advantage of this feature. Do we need to make changes to test utilities to account for new expected values?